pg-extend/src/pg_datum.rs

```diff
@@ -110,6 +112,75 @@ impl From<i64> for PgDatum {
     }
 }
 
+impl TryFromPgDatum for String {
+    fn try_from(datum: PgDatum) -> Result<Self, &'static str> {
+        use std::ffi::CStr;
+        use std::os::raw::c_char;
+
+        if let Some(datum) = datum.0 {
+            let text_val = datum as *const pg_sys::text;
+
+            let cstr = unsafe {
+                let val: *mut c_char = pg_sys::text_to_cstring(text_val);
+                pg_sys::pfree(text_val as *mut _);
+
+                CStr::from_ptr(val)
+            };
+
+            match cstr.to_str() {
+                Ok(s) => Ok(s.to_owned()),
+                Err(_) => Err("datum was not valid utf8")
+            }
+        } else {
+            Err("datum was NULL")
+        }
+    }
+}
+
+impl From<String> for PgDatum {
+    fn from(value: String) -> Self {
+        use std::ffi::CString;
+        use std::os::raw::c_char;
+
+        let cstr = CString::new(value).expect("This shouldn't fail");
+        let ptr: *const c_char = cstr.as_ptr();
+
+        let text = unsafe { pg_sys::cstring_to_text(ptr) };
+
+        PgDatum(Some(text as Datum))
+    }
+}
+
+impl TryFromPgDatum for CString {
+    fn try_from(datum: PgDatum) -> Result<Self, &'static str> {
+        use std::os::raw::c_char;
+
+        if let Some(datum) = datum.0 {
+            let text_val = datum as *const pg_sys::text;
+
+            unsafe {
+                let val: *mut c_char = pg_sys::text_to_cstring(text_val);
+                pg_sys::pfree(text_val as *mut _);
```

Comment:
this should happen after the `CStr::from_ptr(val).to_owned()`. Otherwise you're freeing the memory before it's been converted to a CString.

Reply:
Are you sure? [`text_to_cstring()`](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/varlena.c#L202) allocates new memory for the result, so `pfree` is freeing `text_val` but `val` is new. I'm assuming `val` is freed at the end of the block since it's owned in Rust, but I'm not 100% sure about that.

Reply:
Ah. I confused myself and thought you were freeing ‘val’ here. We done need to free text_val, I’m surprised you’re not seeing a segfault. ‘Const *’  is generally an indication that we don’t own the memory, where as ‘mut *’ implies we do.

Rust doesn’t “take ownership” of ‘val’ in this case, the Cstr wrapping it is borrowing it, in Rust terms. basically, don’t free ‘text_val’, do free ‘val’ after the Cstr to_owned conversion.

I think that’s correct, sorry for the confusion.

Reply:
Just double check this one but I think we're done now